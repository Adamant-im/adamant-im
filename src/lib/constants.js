export const EPOCH = Date.UTC(2017, 8, 2, 17, 0, 0, 0)

export const Transactions = {
  SEND: 0,
  SIGNATURE: 1,
  DELEGATE: 2,
  VOTE: 3,
  MULTI: 4,
  DAPP: 5,
  IN_TRANSFER: 6,
  OUT_TRANSFER: 7,
  CHAT_MESSAGE: 8,
  STATE: 9
}

export const Cryptos = {
  ADM: 'ADM',
  ETH: 'ETH',
  BNB: 'BNB',
  DOGE: 'DOGE',
  BZ: 'BZ'
}

export const ERC20 = Object.freeze([
  Cryptos.BNB,
  Cryptos.BZ
])

export const isErc20 = crypto => ERC20.includes(crypto)

/** Number of decimal places for the different crypto amounts */
export const CryptoAmountPrecision = {
  ADM: 2,
  ETH: 6,
  BNB: 6,
  DOGE: 8,
  BZ: 6
}

/** Fees for the misc ADM operations */
export const Fees = {
  /** Storing a value into the KVS */
  KVS: 0.001,
  /** Transfering tokens */
  ADM_TRANSFER: 0.5,
  NOT_ADM_TRANSFER: 0.001
}

/** Regex for detecting of base64 encoded string */
export const base64regex = /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/

export const Symbols = {
  CLOCK: String.fromCharCode(0x23f0), // ⏰
  HOURGLASS: String.fromCharCode(0x23f3), // ⏳
  CROSS: String.fromCharCode(0x274c) // ❌
}

export const Delegates = {
  ACTIVE_DELEGATES: 101
}

/** Gas value for the ETH transfers */
export const ETH_TRANSFER_GAS = 21000

/** Gas value for the ERC-20 transfers */
export const ERC20_TRANSFER_GAS = ETH_TRANSFER_GAS * 10

export default {
  EPOCH,
  Transactions
}

export const UserPasswordArticleLink = 'https://medium.com/adamant-im/more-convenience-login-to-the-web-messenger-with-user-password-9d48a736dfd8'

export const UserPasswordHashSettings = {
  SALT: 'salt',
  ITERATIONS: 100000,
  KEYLEN: 64,
  DIGEST: 'sha512'
}
