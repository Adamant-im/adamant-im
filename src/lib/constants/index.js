import { Cryptos, CryptosInfo, CryptosOrder } from './cryptos'

export const EPOCH = Date.UTC(2017, 8, 2, 17, 0, 0, 0)

export const Transactions = {
  SEND: 0,
  SIGNATURE: 1,
  DELEGATE: 2,
  VOTE: 3,
  MULTI: 4,
  DAPP: 5,
  IN_TRANSFER: 6,
  OUT_TRANSFER: 7,
  CHAT_MESSAGE: 8,
  STATE: 9
}

/**
 * @see https://github.com/Adamant-im/adamant/wiki/Message-Types
 */
export const MessageType = {
  BASIC_ENCRYPTED_MESSAGE: 1,
  RICH_CONTENT_MESSAGE: 2,
  cryptoTransferMessage: (cryptoSymbol) => `${cryptoSymbol.toLowerCase()}_transaction`
}

export const Rates = {
  USD: 'USD',
  EUR: 'EUR',
  RUB: 'RUB',
  CNY: 'CNY',
  JPY: 'JPY'
}

export const RatesNames = {
  [Rates.USD]: 'USD ($)', // United States Dollar
  [Rates.EUR]: 'EUR (‚Ç¨)', // Euro
  [Rates.RUB]: 'RUB (‚ÇΩ)', // Russian Ruble
  [Rates.CNY]: 'CNY (¬•)', // Chinese Yuan
  [Rates.JPY]: 'JPY (¬•)' // Japanese Yen
}

/** Fees for the misc ADM operations */
export const Fees = {
  /** Storing a value into the KVS */
  KVS: 0.001,
  /** Transferring tokens */
  ADM_TRANSFER: 0.5,
  NOT_ADM_TRANSFER: 0.001
}

/** Regex for detecting of base64 encoded string */
export const base64regex =
  /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/

export const Symbols = {
  CLOCK: String.fromCharCode(0x23f0), // ‚è∞
  HOURGLASS: String.fromCharCode(0x23f3), // ‚è≥
  CROSS: String.fromCharCode(0x274c) // ‚ùå
}

export const Delegates = {
  ACTIVE_DELEGATES: 101
}

export const WelcomeMessage = {
  ADAMANT_BOUNTY: 'ADAMANT Bounty',
  ADAMANT_ICO: 'ADAMANT Tokens'
}

export const BTC_BASED = Object.freeze([Cryptos.DOGE, Cryptos.DASH, Cryptos.BTC])

export const LSK_BASED = Object.freeze([Cryptos.LSK])

export const INSTANT_SEND = Object.freeze([Cryptos.DASH])

// Some cryptos allows to save public data with a Tx
export const ALLOW_TEXT_DATA = Object.freeze([Cryptos.LSK])

export const isErc20 = (crypto) => CryptosInfo[crypto].type === 'ERC20'

export const isEthBased = (crypto) => isErc20(crypto) || crypto === Cryptos.ETH

export const isFeeEstimate = (crypto) => isEthBased(crypto)

export const isBtcBased = (crypto) => BTC_BASED.includes(crypto)

export const isLskBased = (crypto) => LSK_BASED.includes(crypto)

export const isSelfTxAllowed = (crypto) => LSK_BASED.includes(crypto) || crypto === Cryptos.ADM

export const isInstantSendPossible = (crypto) => INSTANT_SEND.includes(crypto)

export const isTextDataAllowed = (crypto) => ALLOW_TEXT_DATA.includes(crypto)

export const RE_LSK_ADDRESS_LEGACY = /^[0-9]{2,21}L$/

/**
 * These gas limit values are used only for estimate fees for ETH & ERC-20 transfers in the Send tokens form,
 * Actual gas limit values are calculated with estimateGas(transactionObject)
 * when each specific transaction is created
 */

/** Gas limit value for the ETH transfers */
export const DEFAULT_ETH_TRANSFER_GAS = CryptosInfo['ETH'].defaultGasLimit
/** Gas limit value for the ERC-20 transfers */
export const DEFAULT_ERC20_TRANSFER_GAS = DEFAULT_ETH_TRANSFER_GAS * 2.4

/** Increase fee multiplier. Used as a checkbox on SendFundsForm */
export const INCREASE_FEE_MULTIPLIER = 2

export { Cryptos, CryptosInfo, CryptosOrder }

export default {
  EPOCH,
  Transactions
}

export const UserPasswordArticleLink =
  'https://medium.com/adamant-im/more-convenience-login-to-the-web-messenger-with-user-password-9d48a736dfd8'

export const UserPasswordHashSettings = {
  SALT: 'salt',
  ITERATIONS: 100000,
  KEYLEN: 64,
  DIGEST: 'sha512'
}

/** Status of ADM or coin transaction */
export const TransactionStatus = {
  CONFIRMED: 'CONFIRMED', // Tx has at least 1 network confirmation
  PENDING: 'PENDING', // We don't know about this Tx anything yet, it may not exist in a blockchain
  REGISTERED: 'REGISTERED', // This Ts is seen on a blockchain, but has 0 confirmations yet
  REJECTED: 'REJECTED', // We unable to find this Tx in a blockchain, or it says it was rejected
  INVALID: 'INVALID', // Wrong sender, recipient, time or amount
  UNKNOWN: 'UNKNOWN' // We don't recognize a cryptocurrency
}

/** Additional status of ADM or coin transaction, which leads to new virtualStatus */
export const TransactionAdditionalStatus = {
  NONE: false,
  INSTANT_SEND: 'instant_send', // Dash InstantSend enabled transaction
  ADM_REGISTERED: 'adm_registered' // ADM tx, registered in a blockchain, but has 0 confirmations yet
}

export const tsIcon = function (status) {
  if (status === TransactionStatus.CONFIRMED) {
    return 'mdi-check'
  } else if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return 'mdi-clock-outline'
  } else if (status === TransactionStatus.REJECTED) {
    return 'mdi-close-circle-outline'
  } else if (status === TransactionStatus.INVALID) {
    return 'mdi-alert-outline'
  } else if (status === TransactionStatus.UNKNOWN) {
    return 'mdi-help-circle-outline'
  }
}

export const tsColor = function (status) {
  if (status === TransactionStatus.REJECTED) {
    return 'red'
  } else if (status === TransactionStatus.INVALID || status === TransactionStatus.UNKNOWN) {
    return 'yellow'
  }
  return ''
}

export const tsUpdatable = function (status, currency) {
  currency = currency.toUpperCase()
  if (currency === Cryptos.ADM) {
    return false
  } else if (status === TransactionStatus.CONFIRMED) {
    return true
  } else if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return false
  } else if (status === TransactionStatus.REJECTED) {
    return true
  } else if (status === TransactionStatus.INVALID) {
    return true
  } else if (status === TransactionStatus.UNKNOWN) {
    return false
  }
}

/**
 * Returns minimal amount that can be transferred for the specified crypto
 * @param {string} crypto crypto
 * @returns {number}
 */
export function getMinAmount(crypto) {
  let amount = CryptosInfo[crypto].minTransferAmount

  if (!amount) {
    const precision = CryptosInfo[crypto].cryptoTransferDecimals
    amount = precision ? Math.pow(10, -precision) : 0
  }

  return amount
}

/**
 * When clicking on replied message inside the chat that animation duration will
 * be applied while scrolling.
 **/
export const SCROLL_TO_REPLIED_MESSAGE_ANIMATION_DURATION = 300 // ms

export const FetchStatus = {
  Idle: 'idle',
  Loading: 'loading',
  Success: 'success',
  Error: 'error'
}

export const REACT_EMOJIS = {
  FACE_WITH_TEARS_OF_JOY: 'üòÇ',
  FIRE: 'üî•',
  GRINNING_FACE_WITH_SMILING_EYES: 'üòÑ',
  THUMB_UP: 'üëç',
  OK_HAND: 'üëå',
  RED_HEART: '‚ù§Ô∏è',
  SLIGHTLY_SMILING_FACE: 'üôÇ',
  THINKING_FACE: 'ü§î',
  WAVING_HAND: 'üëã',
  FOLDED_HANDS: 'üôè',
  FLUSHED_FACE: 'üò≥',
  PARTY_POPPER: 'üéâ'
}
