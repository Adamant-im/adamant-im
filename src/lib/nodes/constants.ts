export type TNodeLabel =
  | 'adm-node'
  | 'eth-node'
  | 'eth-indexer'
  | 'btc-node'
  | 'btc-indexer'
  | 'doge-node'
  | 'doge-indexer'
  | 'dash-node'
  | 'ipfs-node'
  | 'kly-node'
  | 'kly-indexer'
  | 'rates-info'

type KebabToCamelCase<S extends string> = S extends `${infer T}-${infer U}`
  ? `${T}${Capitalize<KebabToCamelCase<U>>}`
  : S

type KebabToPascalCase<S extends string> = Capitalize<KebabToCamelCase<S>>

type CamelToKebabCase<S extends string> = S extends `${infer T}${infer U}`
  ? `${T extends Capitalize<T> ? '-' : ''}${Lowercase<T>}${CamelToKebabCase<U>}`
  : S

type PascalToKebabCase<S extends string> = CamelToKebabCase<Uncapitalize<S>>

type NodeLabels<V extends TNodeLabel = TNodeLabel> = {
  [K in KebabToPascalCase<V>]: Extract<V, PascalToKebabCase<K>>
}

export const NODE_LABELS: NodeLabels = {
  AdmNode: 'adm-node',
  EthNode: 'eth-node',
  EthIndexer: 'eth-indexer',
  BtcNode: 'btc-node',
  BtcIndexer: 'btc-indexer',
  DogeNode: 'doge-node',
  DogeIndexer: 'doge-indexer',
  DashNode: 'dash-node',
  IpfsNode: 'ipfs-node',
  KlyNode: 'kly-node',
  KlyIndexer: 'kly-indexer',
  RatesInfo: 'rates-info'
}
