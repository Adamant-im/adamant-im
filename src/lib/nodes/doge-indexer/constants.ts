export const NB_BLOCKS = 5 // Number of last blocks for fee estimation
