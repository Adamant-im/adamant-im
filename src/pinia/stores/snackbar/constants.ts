export const SNACKBAR_STORE_NAME = 'snackbar' as const
export const SNACKBAR_TIMEOUT = 2000
