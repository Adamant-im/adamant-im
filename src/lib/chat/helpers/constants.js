import { Cryptos } from '@/lib/constants'

/**
 * Maps crypto transfer transaction types to the respective cryptos.
 * E.g.:
 *
 * ```
 * {
 *   'eth_transaction': 'ETH',
 *   'bnb_transaction': 'BNB'
 * }
 * ```
 */
export const KnownCryptos = Object.keys(Cryptos).reduce((map, crypto) => {
  if (crypto !== Cryptos.ADM) {
    const key = `${crypto.toLowerCase()}_transaction`
    map[key] = crypto
  }
  return map
}, {})

/** Cryptos, supported by other clients, but not PWA */
export const UnsupportedCryptos = {
  lsk_transaction: 'LSK'
}
