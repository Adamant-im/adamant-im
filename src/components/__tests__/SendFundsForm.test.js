import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import Vuetify from 'vuetify'
import sinon from 'sinon'

import { Cryptos } from '@/lib/constants'
import SendFundsForm from '@/components/SendFundsForm'
import mockupI18n from './__mocks__/plugins/i18n'

Vue.use(Vuex)
Vue.use(VueI18n)
Vue.use(Vuetify)

// Because Node.js is not supporting Promise.finally.
// In the future, polyfill can be added.
// eslint-disable-next-line
Promise.prototype.finally = Promise.prototype.finally || {
  finally (fn) {
    const onFinally = cb => Promise.resolve(fn()).then(cb)
    return this.then(
      result => onFinally(() => result),
      reason => onFinally(() => { throw reason })
    )
  }
}.finally

// Mock console.error
global.console = {
  ...global.console,
  error: () => {}
}

/** Mockup store helper **/
function mockupStore () {
  // mockup sendTokens action
  function sendTokens ({ admAddress }) {
    return new Promise((resolve, reject) => {
      if (admAddress === 'U111111') {
        return resolve('1') // transactionId
      }

      reject(new Error('Invalid address'))
    })
  }

  const mainModule = () => ({
    state: {
      balance: 10000
    }
  })

  const admModule = () => ({
    state: {
      address: 'U123456'
    },
    getters: {
      fee: () => 0.5
    },
    actions: {
      sendTokens
    },
    namespaced: true
  })

  const ethModule = () => ({
    state: {
      balance: 100
    },
    getters: {
      fee: () => 1
    },
    actions: {
      sendTokens
    },
    namespaced: true
  })

  const bnbModule = () => ({
    state: {
      balance: 1000
    },
    getters: {
      fee: () => 2
    },
    actions: {
      sendTokens
    },
    namespaced: true
  })

  const partnersModule = () => ({
    getters: {
      cryptoAddress: state => (userId, cryptoCurrency) => {
        if (cryptoCurrency === 'ETH') {
          return 'ETH123456'
        } else if (cryptoCurrency === 'BNB') {
          return 'BNB123456'
        }
      },
      displayName: state => partnerId => {
        if (partnerId === 'U111111') {
          return 'Rick'
        } else if (partnerId === 'U222222') {
          return 'Morty'
        }
      }
    },
    actions: {
      fetchAddress () {
        return new Promise((resolve) => {
          resolve('0x111111111111111111')
        })
      }
    },
    namespaced: true
  })

  const chatModule = () => ({
    getters: {
      isPartnerInChatList: state => partnerId => {
        if (partnerId === 'U111111') {
          return true
        }

        return false
      }
    },
    namespaced: true
  })

  return {
    mainModule,
    admModule,
    ethModule,
    bnbModule,
    partnersModule,
    chatModule
  }
}

describe('SendFundsForm', () => {
  let i18n, store, main, adm, eth, bnb, partners, chat, wrapper

  beforeEach(() => {
    const { mainModule, admModule, ethModule, bnbModule, partnersModule, chatModule } = mockupStore()

    main = mainModule()
    adm = admModule()
    eth = ethModule()
    bnb = bnbModule()
    partners = partnersModule()
    chat = chatModule()

    store = new Vuex.Store({
      ...main,
      modules: {
        adm,
        eth,
        bnb,
        partners,
        chat
      }
    })

    i18n = mockupI18n()

    wrapper = shallowMount(SendFundsForm, {
      store,
      i18n
    })
  })

  it('renders the correct markup', () => {
    wrapper = shallowMount(SendFundsForm, {
      store,
      i18n
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('check props', () => {
    wrapper = shallowMount(SendFundsForm, {
      store,
      i18n,
      propsData: {
        cryptoCurrency: 'ADM',
        recipientAddress: 'U111111',
        amountToSend: 100
      }
    })

    expect(wrapper.vm.currency).toBe('ADM')
    expect(wrapper.vm.address).toBe('U111111')
    expect(wrapper.vm.amount).toBe(100)
  })

  /** computed **/
  describe('computed.transferFee', () => {
    it('should return ADM fee', () => {
      wrapper.setProps({
        cryptoCurrency: 'ADM'
      })

      expect(wrapper.vm.transferFee).toBe(0.5)
      expect(wrapper.vm.transferFeeFixed).toBe('0.5')
    })

    it('should return ETH fee', () => {
      wrapper = shallowMount(SendFundsForm, {
        store,
        i18n,
        propsData: {
          cryptoCurrency: 'ETH'
        }
      })

      expect(wrapper.vm.transferFee).toBe(1)
      expect(wrapper.vm.transferFeeFixed).toBe('1')
    })
  })

  describe('computed.finalAmount', () => {
    it('should sum `transferFee` & `amount`', () => {
      wrapper.setData({
        amount: 100
      })

      expect(wrapper.vm.finalAmount).toBe(100 + 0.5)
      expect(wrapper.vm.finalAmountFixed).toBe(String(100 + 0.5))
    })
  })

  describe('computed.balance', () => {
    it('should return ADM balance', () => {
      wrapper = shallowMount(SendFundsForm, {
        store,
        i18n,
        propsData: {
          cryptoCurrency: 'ADM'
        }
      })

      expect(wrapper.vm.balance).toBe(10000)
    })

    it('should return ETH balance', () => {
      wrapper = shallowMount(SendFundsForm, {
        store,
        i18n,
        propsData: {
          cryptoCurrency: 'ETH'
        }
      })

      expect(wrapper.vm.balance).toBe(100)
    })
  })

  describe('computed.maxToTransfer', () => {
    it('should return `balance - fee`', () => {
      expect(wrapper.vm.maxToTransfer).toBe(10000 - 0.5)
    })

    it('should return 0 when `balance < transferFee`', () => {
      main.state.balance = 0 // 0 < 0.5 (ADM_FEE)

      expect(wrapper.vm.maxToTransfer).toBe(0)
    })
  })

  describe('computed.recipientName', () => {
    it('should return `recipientName`', () => {
      wrapper.setData({ address: 'U111111' })

      expect(wrapper.vm.recipientName).toBe('Rick')
    })
  })

  describe('computed.cryptoList', () => {
    it('should return array of available cryptos', () => {
      expect(wrapper.vm.cryptoList).toEqual(Object.keys(Cryptos))
    })
  })

  describe('computed.validationRules', () => {
    it('cryptoAddress: validateAddress', () => {
      const validateAddress = wrapper.vm.validationRules.cryptoAddress[0].bind(wrapper.vm)
      const errorMessage = i18n.t('transfer.error_incorrect_address', { crypto: 'ADM' })

      wrapper.setData({ currency: 'ADM' })

      expect(validateAddress()).toBe(errorMessage)
      expect(validateAddress('U1')).toBe(errorMessage)
      expect(validateAddress('U654321')).toBe(true)
    })

    it('cryptoAddress: validateSameRecipient', () => {
      const validateSameRecipient = wrapper.vm.validationRules.cryptoAddress[1].bind(wrapper.vm)
      const errorMessage = i18n.t('transfer.error_same_recipient')

      expect(validateSameRecipient('U123456')).toBe(errorMessage)
      expect(validateSameRecipient('U654321')).toBe(true)
    })

    it('amount: incorrectAmount', () => {
      const correctAmount = wrapper.vm.validationRules.amount[0].bind(wrapper.vm)
      const errorMessage = i18n.t('transfer.error_incorrect_amount')

      expect(correctAmount('100')).toBe(true)
      expect(correctAmount('0')).toBe(errorMessage)
      expect(correctAmount('-1')).toBe(errorMessage)
      expect(correctAmount('no number')).toBe(errorMessage)
    })

    it('amount: notEnoughTokens', () => {
      const hasEnoughTokens = wrapper.vm.validationRules.amount[1].bind(wrapper.vm)
      const errorMessage = i18n.t('transfer.error_not_enough')

      main.state.balance = 100

      wrapper.setData({ amount: 90 })
      expect(hasEnoughTokens()).toBe(true)

      wrapper.setData({ amount: 99.5 }) // +0.5 ADM transfer fee
      expect(hasEnoughTokens()).toBe(true)

      wrapper.setData({ amount: 100 })
      expect(hasEnoughTokens()).toBe(errorMessage)

      wrapper.setData({ amount: 200 })
      expect(hasEnoughTokens()).toBe(errorMessage)
    })
  })

  describe('computed.confirmMessage', () => {
    it('confirm message without `recipientName`', () => {
      wrapper.setData({
        amount: 1,
        cryptoAddress: 'U333333' // no recipient in `partners` module
      })

      expect(wrapper.vm.confirmMessage).toBe(
        i18n.t('transfer.confirm_message', { amount: 1, address: 'U333333', crypto: 'ADM' })
      )
    })

    it('confirm message with `recipientName`', () => {
      wrapper.setData({
        amount: 1,
        cryptoAddress: 'U111111',
        address: 'U111111'
      })

      expect(wrapper.vm.confirmMessage).toBe(
        i18n.t('transfer.confirm_message_with_name', { amount: 1, name: 'Rick', address: 'U111111', crypto: 'ADM' })
      )
    })
  })

  /** methods **/
  describe('methods.submit', () => {
    const transactionId = 'T1'

    beforeEach(() => {
      wrapper = shallowMount(SendFundsForm, {
        store,
        i18n,
        stubs: {
          vForm: {
            render: (h) => h('div'),
            methods: {
              validate: () => true
            }
          }
        }
      })
    })

    it('should sendFunds and resolve promise', async () => {
      wrapper.setMethods({
        sendFunds: () => Promise.resolve(transactionId),
        pushTransactionToChat: jest.fn()
      })

      const promise = wrapper.vm.submit()

      expect(wrapper.vm.disabledButton).toBe(true)
      expect(wrapper.vm.showSpinner).toBe(true)
      expect(await promise).toBe(undefined)
      expect(wrapper.vm.disabledButton).toBe(false)
      expect(wrapper.vm.showSpinner).toBe(false)
      expect(wrapper.vm.dialog).toBe(false)
      expect(wrapper.emitted('send')).toEqual([[transactionId, 'ADM']])
      expect(wrapper.vm.pushTransactionToChat).not.toHaveBeenCalled()
    })

    it('In-chat Transfer: should sendFunds and push transaction to chat', async () => {
      wrapper.setMethods({
        sendFunds: () => Promise.resolve(transactionId),
        pushTransactionToChat: sinon.spy()
      })
      wrapper.setData({ address: 'U111111' })
      wrapper.setData({ cryptoAddress: 'U111111' })

      await wrapper.vm.submit()

      expect(wrapper.vm.pushTransactionToChat.args).toEqual([
        [transactionId, 'U111111']
      ])
    })

    it('Direct Transfer: should sendFunds', async () => {
      wrapper.setMethods({
        sendFunds: () => Promise.resolve(transactionId),
        pushTransactionToChat: sinon.spy()
      })
      wrapper.setData({ cryptoAddress: 'U111111' })

      await wrapper.vm.submit()

      expect(wrapper.vm.pushTransactionToChat.args).toEqual([])
    })

    it('should emit error when sendFunds rejected', async () => {
      wrapper.setMethods({
        sendFunds: () => Promise.reject(new Error('No hash'))
      })

      await wrapper.vm.submit()

      expect(wrapper.emitted('error')).toEqual([['No hash']])
    })
  })

  describe('methods.sendFunds', () => {
    it('should send ADM with message', () => {})
    it('should send ADM without message', () => {})
    it('should send OTHER crypto', () => {})
  })

  describe('methods.pushTransactionToChat', () => {
    it('should push ADM transaction', () => {})
    it('should push OTHER crypto transaction', () => {})
  })

  describe('methods.fetchUserCryptoAddress', () => {
    it('should resolve with ADM address', () => {})
    it('should fetch and resolve OTHER crypto address', () => {})
  })

  describe('methods.validateNaturalUnits', () => {
    it('should validate natural units of `amount`', () => {
      const validateNaturalUnits = wrapper.vm.validateNaturalUnits.bind(wrapper.vm)

      expect(validateNaturalUnits(0.00000001, 'ADM')).toBe(true)
      expect(validateNaturalUnits(0.000000001, 'ADM')).toBe(false)
      expect(validateNaturalUnits(1, 'ADM')).toBe(true)
    })
  })
})
