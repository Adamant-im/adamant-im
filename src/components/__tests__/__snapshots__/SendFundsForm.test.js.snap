// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendFundsForm renders the correct markup 1`] = `
<div
  class="send-funds-form"
>
  <v-form-stub
    value="true"
  >
    <v-select-stub
      class="a-input"
      items="ADM,ETH,BZ,KCS,DOGE,DASH,BNB,USDS"
    />
     
    <v-text-field-stub
      class="a-input"
      type="text"
    >
      <template>
        <span
          class="font-weight-medium"
        >
          
          To Address
        
        </span>
      </template>
    </v-text-field-stub>
     
    <v-text-field-stub
      class="a-input"
      max="9999.5"
      min="0.01"
      step="0.01"
      type="number"
    >
      <template>
        <span
          class="font-weight-medium"
        >
          Amount to send
        </span>
         
        <span
          class="body-1"
        >
          
          (max: 9999.5 ADM)
        
        </span>
      </template>
    </v-text-field-stub>
     
    <v-text-field-stub
      class="a-input"
      disabled=""
      label="Transfer fee"
      value="0.5 ADM"
    />
     
    <v-text-field-stub
      class="a-input"
      disabled=""
      label="Amount including a transfer fee"
      value="0.5 ADM"
    />
     
    <!---->
     
    <div
      class="text-xs-center"
    >
      <v-btn-stub
        activeclass="v-btn--active"
        class="a-btn-primary send-funds-form__button"
        tag="button"
        type="button"
      >
        
        Send
      
      </v-btn-stub>
    </div>
  </v-form-stub>
   
  <v-dialog-stub
    maxwidth="none"
    origin="center center"
    transition="dialog-transition"
    width="500"
  >
    <v-card-stub>
      <v-card-title-stub
        class="a-text-header"
      >
        Transfer confirmation
      </v-card-title-stub>
       
      <v-divider-stub
        class="a-divider"
      />
       
      <v-card-text-stub
        class="a-text-regular-enlarged"
      />
       
      <v-card-actions-stub>
        <v-spacer-stub />
         
        <v-btn-stub
          activeclass="v-btn--active"
          class="a-btn-regular"
          flat="true"
          tag="button"
          type="button"
        >
          
          Cancel
        
        </v-btn-stub>
         
        <v-btn-stub
          activeclass="v-btn--active"
          class="a-btn-regular"
          flat="true"
          tag="button"
          type="button"
        >
          <v-progress-circular-stub
            class="mr-3"
            color="primary"
            indeterminate="true"
            rotate="0"
            size="24"
            style="display: none;"
            value="0"
            width="4"
          />
          
          Confirm
        
        </v-btn-stub>
      </v-card-actions-stub>
    </v-card-stub>
  </v-dialog-stub>
</div>
`;
